bin/generate-npm-project-based-fixtures.rb

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require "fileutils"
+require "byebug"
+require "open3"
+
+# Generate project based fixtures for npm and yarn
+class ProjectBasedFixtureGenerator
+  def run
+    FileUtils.mkdir_p(npm6_project_dir)
+    FileUtils.mkdir_p(yarn_project_dir)
+    _, _, status = Open3.capture3("npm install --global npm@6")
+    abort("failed to install npm@6 globally") unless status.success?
+
+    _, _, status = Open3.capture3("yarn upgrade v1.22.5")
+    abort("failed to install yarn v1.22.5") unless status.success?
+
+    npm_package_files.each do |filename|
+      build_npm_project(filename)
+      build_yarn_project(filename)
+    end
+  end
+
+  private
+
+  def build_npm_project(filename)
+    project_name = File.basename(filename, ".json")
+    project_dir = FileUtils.mkdir_p(File.join(npm6_project_dir, project_name)).last
+    FileUtils.copy(filename, File.join(project_dir, "package.json"))
+    Dir.chdir(project_dir) do
+      _, _, status = Open3.capture3("npm install")
```

Comment:
Oh slight concern, I know we have some test fixtures where the package.json wouldn't generate the same lockfile that's currently being tested. I've committed examples where I had to pin the version in `package.json` to get a particular lockfile but then needed the version to not be pinned in `package.json` so changed the requirement after the lockfile was generated 😬

This should result in broken tests where things have changed so should be ok to continue with this and fix up the broken tests when we encounter them. Seems like a good time to try and change the manifest to generate the correct lockfile from the get go where possible.

Reply:
Yeah I touched on this slightly in the PR description, there will probably be more cases where this doesn't fully cover what we need it to, I think the only thing we can hope for with this PR is to reduce noise in future PRs where we actually update the test. We'll still need to verify that the projects are set up correctly for what we're trying to test.

Reply:
Yeah I reckon we should get this in and go from there 👍